### ISSUE TITLE
[Question]  Can  PythonPackageVersionEntity have multiple same value

### ISSUE THREAD

**Issue by harshad16** [thoth-station/storages#2729]
The database shows multiple entry of the same unique tuple
![Screenshot from 2022-10-27 15-57-28](https://user-images.githubusercontent.com/14028058/198386163-1c520e67-08ae-4dbb-8e31-c489e18f1535.png)
even though we have set the unique constriat https://github.com/thoth-station/storages/blob/a69733ccfa0cd3d2afd5f00e4a51fb264f692cb6/thoth/storages/graph/models.py#L203

The python_package_index_id, seems to be the differentiating factor, which somehow hasn't been included.

**Comment by mayaCostantini** [comment]
Which is the postgres image used to run the database? Is the version < 15?
Postgres versions before 15 seems to consider NULL values as distinct, which would impose to use something like `postgresql_where=python_package_index_id.isnot(None)` in the `python_package_version_entity_idx` index:

https://github.com/thoth-station/storages/blob/a69733ccfa0cd3d2afd5f00e4a51fb264f692cb6/thoth/storages/graph/models.py#L204-L210

```
      Index(
          "python_package_version_entity_idx",
          "package_name",
          "package_version",
          "python_package_index_id",
          postgresql_where=python_package_index_id.isnot(None),
          unique=True,
          ),
```

I have not tested this solution though as I still have issues with podman. I will sort this out and get back to you if that works.

**Comment by harshad16** [comment]
we use Postgresql 10 in our systems.
I wasn't aware that the older version considered NULL as distinct.

Maybe it is time to upgrade to a newer version of PostgreSQL and clean the database a little.
we should identify what are the change that we would have to make for such upgrade

**Comment by mayaCostantini** [comment]
Agree, I can assign myself to this issue and start working on it when I solve my testing environment issues. What would you say is the priority of this?

**Comment by mayaCostantini** [comment]
Also, this might be related to https://github.com/thoth-station/storages/issues/2189

**Comment by harshad16** [comment]
#2189 could be related
It would be a priority as I feel this might affect the data ingestion.

**Comment by harshad16** [comment]
Questions:

- Isn't there a get_or_create provided by sqlalchemy itself ? Catching integrity error + session.rollback looks like the job of a context manager
